web.go

```diff
@@ -0,0 +1,190 @@
+package gorillaz
+
+import (
+	"context"
+	"github.com/gorilla/websocket"
+	"github.com/skysoft-atm/gorillaz/mux"
+	"go.uber.org/zap"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"time"
+)
+
+// serves a file as an http response
+func ServeFileFunc(file string) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "text/html; charset=UTF-8")
+		b, err := ioutil.ReadFile(filepath.Clean(file))
+		if err != nil {
+			w.WriteHeader(http.StatusNotFound)
+		} else {
+			w.WriteHeader(http.StatusOK)
+			_, err = w.Write(b)
+			if err != nil {
+				Log.Debug("Could not write response on http connection", zap.Error(err))
+			}
+		}
+	}
+}
+
+type WebsocketConfig struct {
+	WriteWait  time.Duration // Time allowed to write a message to the peer.
+	PongWait   time.Duration // Time allowed to read the next pong message from the peer.
+	PingPeriod time.Duration // Send pings to peer with this period. Must be less than pongWait.
+}
+
+type WebsocketOption func(*WebsocketConfig)
+
+// Websocket message
+type WsMessage struct {
+	MessageType int
+	Data        []byte
+}
+
+// Upgrades the http request to a websocket connection
+func UpgradeToWebsocketWithContext(rw http.ResponseWriter, req *http.Request, opts ...WebsocketOption) (chan<- *WsMessage, context.Context, error) {
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+	}
+	c := WebsocketConfig{
+		WriteWait:  10 * time.Second,
+		PongWait:   60 * time.Second,
```

Comment:
Where is it used?

Reply:
this can be used by the servers using this API to publish data over websocket, in order to configure different timeouts / intervals